### PR TITLE
javascript: actually a typescript implementation

### PR DESCRIPTION
**Pull request by progrium**

### CODE REVIEW

File: javascript/api.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+	write(buffer: Uint8Array): Promise<number>;
+	close(): Promise<void>;
+}
+
+export interface ISession {
+	open(): Promise<IChannel>;
+	accept(): Promise<IChannel>;
+    close(): Promise<void>;
+
+}
```
Comment (mgood) [inline javascript/api.ts:13]: The indentation here is inconsistent, looks like a mix of tabs and spaces. I'm seeing that in the other files too.

Comment (progrium) [reply javascript/api.ts:13]: ugh, kill me. i will see if there is a formatter plugin for vscode

Comment (mgood) [reply javascript/api.ts:13]: Yeah, I'm not super-familiar with the formatting options myself. IIRC we were using `TSLint` at work, but that's kind of deprecated now. The VSCode built-in TypeScript language support does seem to offer formatting though, so I ran through converting tabs to spaces and running the formatter with this PR: #7.

Comment (mgood) [reply javascript/api.ts:13]: I also had Prettier installed from some previous stuff, but figured sticking with the built-in one is simpler if it works well enough. It changes some of the inline spacing around operators, and removed some of the alignment of fields. There may be options for some of those things, but I haven't explored that yet.

File: javascript/api.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
+export interface IConn {
+	read(len: number): Promise<Uint8Array|undefined>;
```
Comment (mgood) [inline javascript/api.ts:3]: is this `undefined` when you reach the EOF?

File: javascript/channel.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
+
+import * as util from "./util.ts";
```
Comment (mgood) [inline javascript/channel.ts:2]: VSCode tells me:
```
An import path cannot end with a '.ts' extension. Consider importing './util' instead.
```
Should these be changed, or is there a setting I should update for this import convention?

File: javascript/api.ts (hunk 1 above)
Comment (progrium) [reply javascript/api.ts]: 👍

File: javascript/api.ts (hunk 2 above)
Comment (progrium) [reply javascript/api.ts]: i believe so

**Pull request merged**